File: client/py/synnax/channel/registry.py
#  Copyright 2023 Synnax Labs, Inc.
#
#  Use of this software is governed by the Business Source License included in the file
#  licenses/BSL.txt.
#
#  As of the Change Date specified in that file, in accordance with the Business Source
#  License, use of this software will be governed by the Apache License, Version 2.0,
#  included in the file licenses/APL.txt.
#
#  Use of this software is governed by the Business Source License included in the file
#  licenses/BSL.txt.
#
#  As of the Change Date specified in that file, in accordance with the Business Source
#  License, use of this software will be governed by the Apache License, Version 2.0,
#  included in the file licenses/APL.txt.

from .payload import ChannelPayload
from .retrieve import ChannelRetriever


class ChannelRegistry:
    retriever: ChannelRetriever
    channels: dict[str, ChannelPayload]

    def __init__(self, retriever: ChannelRetriever, channels=None) -> None:
        if channels is None:
            channels = list()
        self.retriever = retriever
        self.channels = {ch.name: ch for ch in channels}

    def get(self, key: str) -> ChannelPayload | None:
        record = self.channels.get(key, None)
        if record is None:
            record = self.retriever.retrieve(key=key)
            self.channels[key] = record
        return record

    def get_n(self, keys: list[str]) -> list[ChannelPayload]:
        results = list()
        retrieve_keys = list()
        for key in keys:
            record = self.channels.get(key, None)
            if record is not None:
                results.append(record)
            retrieve_keys.append(key)
        if retrieve_keys:
            results.extend(self.retriever.filter(keys=retrieve_keys))
        return results

File: client/ts/src/lib/channel/registry.ts
// Copyright 2022 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { ChannelPayload } from './payload';
import Retriever from './retriever';

export default class Registry {
	private readonly retriever: Retriever;
	private readonly channels: Map<string, ChannelPayload>;

	constructor(retriever: Retriever) {
		this.retriever = retriever;
		this.channels = new Map();
	}

	async get(key: string): Promise<ChannelPayload> {
		let channel = this.channels.get(key);
		if (channel === undefined) {
			channel = await this.retriever.retrieve({ key });
			this.channels.set(key, channel);
		}
		return channel;
	}

	async getN(...keys: string[]): Promise<ChannelPayload[]> {
		const results: ChannelPayload[] = [];
		const retrieveKeys: string[] = [];
		keys.forEach((key) => {
			const channel = this.channels.get(key);
			if (channel === undefined) retrieveKeys.push(key);
			else results.push(channel);
		});
		if (retrieveKeys.length > 0) {
			const channels = await this.retriever.filter({ keys: retrieveKeys });
			channels.forEach((channel) => {
				this.channels.set(channel.key as string, channel);
				results.push(channel);
			});
		}
		return results;
	}
}
